src/transformStateWithClones.js

```diff
@@ -7,7 +7,28 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const transformedState = structuredClone(state);
+  const stateHistory = [];
+
+  for (let i = 0; i < actions.length; i++) {
+    if (actions[i].type === 'addProperties') {
+      for (const key in actions[i].extraData) {
+        transformedState[key] = actions[i].extraData[key];
+      }
+    } else if (actions[i].type === 'removeProperties') {
+      for (const key of actions[i].keysToRemove) {
+        delete transformedState[key];
+      }
+    } else {
+      for (const key in transformedState) {
+        delete transformedState[key];
+      }
+    }
```

Comment:
A `switch` statement would be more appropriate here as per checklist item 2 since you have a limited and known set of action types to handle.